connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/ProtobufDeserializer.scala

```diff
@@ -0,0 +1,373 @@
+
+      case (FLOAT, ArrayType(FloatType, containsNull)) =>
+        newArrayWriter(protoType, protoPath,
+          catalystPath, FloatType, containsNull)
+
+      case (DOUBLE, DoubleType) => (updater, ordinal, value) =>
+        updater.setDouble(ordinal, value.asInstanceOf[Double])
+
+      case (DOUBLE, ArrayType(DoubleType, containsNull)) =>
+        newArrayWriter(protoType, protoPath,
+          catalystPath, DoubleType, containsNull)
+
+      case (STRING, StringType) => (updater, ordinal, value) =>
+        val str = value match {
+          case s: String => UTF8String.fromString(s)
+        }
+        updater.set(ordinal, str)
+
+      case (STRING, ArrayType(StringType, containsNull)) =>
+        newArrayWriter(protoType, protoPath,
+          catalystPath, StringType, containsNull)
+
+      case (BYTE_STRING, BinaryType) => (updater, ordinal, value) =>
+        val byte_array = value match {
+          case s: ByteString => s.toByteArray
+          case _ => throw new Exception("Invalid ByteString format")
+        }
+        updater.set(ordinal, byte_array)
+
+      case (BYTE_STRING, ArrayType(BinaryType, containsNull)) =>
+        newArrayWriter(protoType, protoPath,
+          catalystPath, BinaryType, containsNull)
+
+      case (MESSAGE, MapType(keyType, valueType, valueContainsNull)) =>
+        newMapWriter(protoType, protoPath, catalystPath,
+          keyType, valueType, valueContainsNull)
+
+      case (MESSAGE, st: StructType) =>
+        val writeRecord = getRecordWriter(protoType.getMessageType, st, protoPath,
+          catalystPath, applyFilters = _ => false)
+        (updater, ordinal, value) =>
+          val row = new SpecificInternalRow(st)
+          writeRecord(new RowUpdater(row), value.asInstanceOf[DynamicMessage])
+          updater.set(ordinal, row)
+
+      case (MESSAGE, ArrayType(st: StructType, containsNull)) =>
+        newArrayWriter(protoType, protoPath,
+          catalystPath, st, containsNull)
+
+      case (ENUM, StringType) => (updater, ordinal, value) =>
+        updater.set(ordinal, UTF8String.fromString(value.toString))
+
+      case (ENUM, ArrayType(StringType, containsNull)) =>
+        newArrayWriter(protoType, protoPath,
+          catalystPath, StringType, containsNull)
+
+      case _ => throw new IncompatibleSchemaException(incompatibleMsg)
+    }
+  }
+
+
+  private def getRecordWriter(
+                               protoType: Descriptor,
+                               catalystType: StructType,
+                               protoPath: Seq[String],
+                               catalystPath: Seq[String],
+                               applyFilters: Int => Boolean):
+  (CatalystDataUpdater, DynamicMessage) => Boolean = {
+
+    val protoSchemaHelper = new ProtobufUtils.ProtoSchemaHelper(
+      protoType, catalystType, protoPath, catalystPath, positionalFieldMatch)
+
+   // protoSchemaHelper.validateNoExtraCatalystFields(ignoreNullable = true)
```

Comment:
could you add comment about this? Do we plan to add this check later?

Reply:
In the schema evolution case e.g., new fields are added, you want to allow extra catalyst fields when you are reading a message from the old schema. That's why this is commented out. There is a test case old Producer -> New Consumer which simulates this

Reply:
There is only one protobuf schema here. Also note that this code runs only once, not for each input message. 
The check is just to ensure that catalystType does not have any fields that are not in Protobuf schema. 

> allow extra catalyst fields when you are reading a message from the old schema.

What does it mean? If there are fields that our prototobuf version does not know (say old version had an extra field that is ignored now), there is noway to know anything about it. We don't know what it was called.

Reply:
If the incoming message has more fields than what we have in the schema, the fields will be ignored. If the schema has more fields than the incoming message, we want to allow that as part of evolution and we are having a discussion below on that on what should be set for the new field. This function will remain uncommented to allow schema evolution. Let me know if there is some misunderstanding

Reply:
I think there is a bit of a confusion. We could do a video call to meet and sync up of couple of things. Meeting will be good anyway.

Reply:
Okay, afternoon looks good to me. We can discuss offline on timings.

Reply:
Sg. We don't need to address this now. Please leave TODO comment to revisit.

Reply:
added a comment to revisit.